server/src/dependency/heed.rs

```diff
@@ -0,0 +1,189 @@
+}
+
+pub fn memory() -> SharedStorage {
+    db()
+}
+
+pub fn block_repository() -> impl BlockRepository<Storage = SharedStorage> + Send + Sync + 'static {
+    block::HeedBlockRepository
+}
+
+pub fn state() -> impl State + Send + Sync + 'static {
+    state::HeedState::new(std::sync::Arc::new(trie::HeedEthTrieDb::new(db())))
+}
+
+pub fn state_query(genesis_config: &GenesisConfig) -> StateQueries {
+    state::HeedStateQueries::from_genesis(db(), genesis_config.initial_state_root)
+}
+
+pub fn on_tx_batch<
+    A: State,
+    B: NewPayloadId,
+    C: BlockHash,
+    D: BlockRepository<Storage = SharedStorage>,
+    E: BaseGasFee,
+    F: CreateL1GasFee,
+    G: CreateL2GasFee,
+    H: BaseTokenAccounts,
+>() -> OnTxBatch<A, B, C, D, E, F, G, H> {
+    Box::new(|| {
+        Box::new(|state| {
+            state
+                .state_queries()
+                .push_state_root(state.state().state_root())
+                .unwrap()
+        })
+    })
+}
+
+pub fn on_tx<
+    A: State,
+    B: NewPayloadId,
+    C: BlockHash,
+    D: BlockRepository<Storage = SharedStorage>,
+    E: BaseGasFee,
+    F: CreateL1GasFee,
+    G: CreateL2GasFee,
+    H: BaseTokenAccounts,
+>() -> OnTx<A, B, C, D, E, F, G, H> {
+    StateActor::on_tx_noop()
+}
+
+pub fn on_payload<
+    A: State,
+    B: NewPayloadId,
+    C: BlockHash,
+    D: BlockRepository<Storage = SharedStorage>,
+    E: BaseGasFee,
+    F: CreateL1GasFee,
+    G: CreateL2GasFee,
+    H: BaseTokenAccounts,
+>() -> OnPayload<A, B, C, D, E, F, G, H> {
+    Box::new(|| {
+        Box::new(|state, id, hash| state.payload_queries().add_block_hash(id, hash).unwrap())
+    })
+}
+
+pub fn transaction_repository() -> TransactionRepository {
+    transaction::HeedTransactionRepository
+}
+
+pub fn transaction_queries() -> TransactionQueries {
+    transaction::HeedTransactionQueries
+}
+
+pub fn receipt_repository() -> ReceiptRepository {
+    receipt::HeedReceiptRepository
+}
+
+pub fn receipt_queries() -> ReceiptQueries {
+    receipt::HeedReceiptQueries
+}
+
+pub fn receipt_memory() -> ReceiptStorage {
+    db()
+}
+
+pub fn block_queries() -> BlockQueries {
+    block::HeedBlockQueries
+}
+
+pub fn payload_queries() -> PayloadQueries {
+    payload::HeedPayloadQueries::new(db())
+}
+
+lazy_static::lazy_static! {
+    static ref Database: moved_storage_heed::Env = {
+        create_db()
+    };
+}
+
+fn db() -> &'static moved_storage_heed::Env {
+    &Database
+}
+
+fn create_db() -> moved_storage_heed::Env {
+    assert_eq!(moved_storage_heed::DATABASES.len(), 9);
+
+    let path = "db";
+
+    if std::fs::exists(path).unwrap() {
+        std::fs::remove_dir_all(path)
+            .expect("Removing non-empty database directory should succeed");
+    }
+    std::fs::create_dir(path).unwrap();
+
+    let env = unsafe {
+        EnvOpenOptions::new()
+            .max_dbs(moved_storage_heed::DATABASES.len() as u32)
+            .map_size(1024 * 1024 * 1024 * 1024) // 1 TiB
```

Comment:
Does it pre-allocate the data file, meaning do we need a server over 1 TiB storage?

Reply:
No, this is an upper limit of the database directory size on disk.

Internally, it means the size of the virtual address space.